### PR TITLE
Using poll instead of select in nodeShareInputScan

### PR DESCRIPTION
**Pull request by interma**
Current 6X ShareInputScan using FIFO as the notification mechanism between processes and using `select()` to detect whether the specific fd can read.
But `select()` has the "**max fd number**" limit, we can see the check logic (65535) in `MPP_FD_SET()`:
```
#define MPP_FD_SET(fd, set) do{ \
	if (fd > 65535) \
	elog(FATAL,"Internal error:  Using fd > 65535 in MPP_FD_SET"); \
	((set)->__fds_bits[MPP_FD_WORD(fd)] = ((set)->__fds_bits[MPP_FD_WORD(fd)]) | MPP_FD_BIT(fd)); \
    } while(0)
```

One error scene example is 
```
 Sequence  (cost=0.00..9854193.27 rows=149 width=205) (actual time=19477.963..20583.114 rows=456618 loops=1)
   ->  Shared Scan (share slice:id 0:0)  (cost=0.00..431.00 rows=1 width=1) (actual time=15.110..15.114 rows=1 loops=1)
         ->  Materialize  (cost=0.00..431.00 rows=1 width=1) (actual time=15.080..15.080 rows=0 loops=1)
               ->  Limit  (cost=0.00..431.00 rows=1 width=7) (actual time=0.268..0.301 rows=1 loops=1)
                     ->  Gather Motion 816:1  (slice83; segments: 816)  (cost=0.00..431.00 rows=1 width=7) (actual time=0.265..0.265 rows=1 loops=1)
                           ->  GroupAggregate  (cost=0.00..431.00 rows=1 width=7) (actual time=2893.051..2893.051 rows=1 loops=1)
...
```
SharedInputScan can run on QD in some plans, that means its fd number is very bigger than normal QE (QD need to access many files and socks, especially in a big cluster). When the number exceeds 65535, an error "FATAL: Internal error: Using fd > 65535 in MPP_FD_SET" happens.

Using `poll()` can avoid this issue: indeed, should use `poll()` instead of `select()` in every place.
https://man7.org/linux/man-pages/man2/select.2.html#DESCRIPTION

Note: master doesn't have this issue, it uses condition variable as the notification mechanism (so no select/poll are needed).

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

**Comment by interma** [comment]
and no new test is needed, the previous tests: gp_aggregates and shared_scan can cover this PR.

**Comment by ashwinstar** [comment]
Understand for functional aspect we don't need any new tests as this code change as is getting tested.

Though the scenario for which this change is being made - QDs fd number can be very big (> 65535) and system performs fine, how are we confident that works fine? Like even if this code path is no more a problem, nothing else in system is limiting that aspect? Ideas how we can validate and confidently convey the scenario works fine? Anyway to simulate and automate the testing for that?

**Comment by interma** [comment]
Thanks Ashwin.
I will try to test it manually first, then consider writing a proper test case.

**Comment by interma** [comment]
Simulate a similar environment and test it works find:
```
638		create_tmp_fifo(pctxt->lkname_ready);
(gdb) n
639		pctxt->readyfd = open(pctxt->lkname_ready, O_RDWR, 0600);
(gdb) n
640		if(pctxt->readyfd < 0)
(gdb) p pctxt->readyfd
$2 = 70011 // > 65535
```

And add a faultinject test case.

File: src/test/regress/expected/shared_scan.out [hunk 1]
```diff
@@ -247,3 +250,36 @@ where
...
+   14 |     1 |     2
+   16 |     1 |     2
+   18 |     1 |     2
+(10 rows)
+
+select gp_inject_fault('inject_many_fds_for_shareinputscan', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
+ gp_inject_fault 
+-----------------
+ Success:
+(1 row)
```
Comment (zxuejing) [inline src/test/regress/expected/shared_scan.out:284]: rm -rf /tmp/_gpdb_fault_inject_tmp_dir/

File: src/backend/executor/nodeShareInputScan.c [hunk 2]
```diff
@@ -547,6 +554,39 @@ static int retry_write(int fd, char *buf, int wsize)
...
+ **/
+static void fi_create_many_fds(int *fds, char *file_prefix, int num)
+{
+	for (int i = 0; i < num; i++)
+	{
+		char filepath[1024];
+		snprintf(filepath, sizeof(filepath), "%s/si_%d", file_prefix, i);
+		fds[i] = open(filepath, O_RDWR | O_CREAT, 0666);
+	}
+	if (fds[num-1] > 0)
```
[the remark this answers is not in the document]
Comment (interma) [reply src/backend/executor/nodeShareInputScan.c:569]: yes, this situation usually happened on test env (with small "open files" in ulimit).

File: src/test/regress/expected/shared_scan.out (hunk 1 above)
Comment (interma) [reply src/test/regress/expected/shared_scan.out:284]: thanks, I will fix it.

**Review by zxuejing — APPROVED** [review]

**Pull request merged**